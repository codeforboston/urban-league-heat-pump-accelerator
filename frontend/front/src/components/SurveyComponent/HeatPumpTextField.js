import { Controller, useController } from "react-hook-form";
import React, { useMemo } from "react";

import { FormControl, TextField, FormLabel } from "@mui/material";

export const HeatPumpTextField = ({
  name,
  control,
  label,
  disabled,
  required,
  type,
  readOnly,
  disableFancyLabel,
}) => {
  const { formState } = useController({ name, control });

  const inputType = useMemo(
    // zipcode is not a real html input type
    () => (!type || type === "zipcode" ? "text" : type),
    [type]
  );

  const rules = useMemo(() => {
    const ruleList = {};

    if (required) {
      ruleList.required = { value: true, message: "This field is required!" };
    }

    // whether or not email is automatically validated seems to depend on the browser
    if (type === "email") {
      ruleList.pattern = {
        value: /^[\w\-.]+@([\w-]+\.)+[\w-]{2,}$/,
        message: "Invalid email address.",
      };
    }

    if (type === "zipcode") {
      ruleList.pattern = {
        value: /\d{5}/,
        message: "Invalid ZIP code.",
      };
    }
  }, [type, required]);

  return (
    <Controller
      name={name}
      control={control}
      rules={rules}
      render={({ field }) => (
        <FormControl fullWidth>
          {disableFancyLabel && (
            <FormLabel
              id={`${name}-textField-label`}
              htmlFor={`${name}-textField`}
            >
              {label}
            </FormLabel>
          )}
          <TextField
            data-testid={name}
            label={disableFancyLabel ? undefined : label}
            variant="standard"
            error={!!formState.errors[name]}
            helperText={
              !!formState.errors[name] && formState.errors[name].message
            }
            disabled={disabled}
            type={inputType}
            inputProps={{ readOnly, id: `${name}-textField` }}
            {...field}
          />
        </FormControl>
      )}
    />
  );
};
