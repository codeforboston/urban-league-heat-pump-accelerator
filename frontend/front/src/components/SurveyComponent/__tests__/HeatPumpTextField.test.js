import { render } from "@testing-library/react";
import React from "react";
import { useForm } from "react-hook-form";
import { HeatPumpTextField } from "../HeatPumpTextField";

const DEFAULT_FORM = { 1: "" };

const DEFAULT_FIELD = {
  id: 1,
  display_order: 1,
  text: "Enter some text!",
  response_type: "text",
};

const FormWrapper = ({ q }) => {
  const { control } = useForm({ defaultValues: DEFAULT_FORM });
  return (
    <HeatPumpTextField
      key={`q${q.id}`}
      control={control}
      name={`${q.id}`}
      label={q.question}
      {...q}
    />
  );
};

describe("HeatPumpTextField", () => {
  it("renders normally with default props", () => {
    const { container } = render(<FormWrapper q={DEFAULT_FIELD} />);

    expect(container).toMatchSnapshot();
  });

  it("renders as expected when disableFancyLabel=true", () => {
    const { container } = render(
      <FormWrapper q={{ ...DEFAULT_FIELD, disableFancyLabel: true }} />
    );

    expect(container).toMatchSnapshot();
  });
});
