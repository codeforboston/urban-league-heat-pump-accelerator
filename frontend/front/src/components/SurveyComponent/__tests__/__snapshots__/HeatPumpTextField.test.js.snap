// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeatPumpTextField renders as expected when disableFancyLabel=true 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiFormLabel-colorPrimary css-u4tvz2-MuiFormLabel-root"
      for="1-textField"
      id="1-textField-label"
    >
      Enter some text!
    </label>
    <div
      class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      data-testid="1"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-v4u5dn-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id="1-textField"
          name="1"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`HeatPumpTextField renders normally with default props 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <div
      class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      data-testid="1"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r0:"
        id=":r0:-label"
      >
        Enter some text!
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-v4u5dn-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id="1-textField"
          name="1"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
